providers/openfeature-provider-flagd/src/openfeature/contrib/provider/flagd/resolvers/process/custom_ops.py

```diff
@@ -0,0 +1,115 @@
+
+    seed = data.get("$flagd", {}).get("flagKey", "")
+    hash_key = seed + bucket_by
+    hash_ratio = abs(mmh3.hash(hash_key)) / (2**31 - 1)
+    bucket = int(hash_ratio * 100)
+
+    for arg in args:
+        if (
+            not isinstance(arg, (tuple, list))
+            or len(arg) != 2
+            or not isinstance(arg[0], str)
+            or not isinstance(arg[1], int)
+        ):
+            raise ValueError("Fractional variant weights must be (str, int) tuple")
+    variant_weights: tuple[tuple[str, int]] = args  # type: ignore[assignment]
+
+    range_end = 0
+    for variant, weight in variant_weights:
+        range_end += weight
+        if bucket < range_end:
+            return variant
+
+    return None
+
+
+def starts_with(data: dict, *args: JsonLogicArg) -> bool:
+    def f(s1: str, s2: str) -> bool:
+        return s1.startswith(s2)
+
+    return string_comp(f, data, *args)
+
+
+def ends_with(data: dict, *args: JsonLogicArg) -> bool:
+    def f(s1: str, s2: str) -> bool:
+        return s1.endswith(s2)
+
+    return string_comp(f, data, *args)
+
+
+def string_comp(
+    comparator: typing.Callable[[str, str], bool], data: dict, *args: JsonLogicArg
+) -> bool:
+    if not args:
+        raise ValueError("No arguments provided to string_comp operator.")
+    if len(args) != 2:
+        raise ValueError("Exactly 2 args expected for string_comp operator.")
+    arg1, arg2 = args
+    if not isinstance(arg1, str):
+        logging.debug(f"incorrect argument for first argument, expected string: {arg1}")
+        return False
+    if not isinstance(arg2, str):
+        logging.debug(
+            f"incorrect argument for second argument, expected string: {arg2}"
+        )
+        return False
+
+    return comparator(arg1, arg2)
+
+
+def sem_ver(data: dict, *args: JsonLogicArg) -> bool:  # noqa: C901
+    if not args:
+        raise ValueError("No arguments provided to sem_ver operator.")
+    if len(args) != 3:
+        raise ValueError("Exactly 3 args expected for sem_ver operator.")
+
+    arg1, op, arg2 = args
+
+    try:
+        v1 = semver.Version.parse(arg1)
+        v2 = semver.Version.parse(arg2)
+    except ValueError as e:
+        logging.exception(e)
+        return False
+
+    if op == "=":
+        return v1 == v2  # type: ignore[no-any-return]
+    elif op == "!=":
+        return v1 != v2  # type: ignore[no-any-return]
+    elif op == "<":
+        return v1 < v2  # type: ignore[no-any-return]
+    elif op == "<=":
+        return v1 <= v2  # type: ignore[no-any-return]
+    elif op == ">":
+        return v1 > v2  # type: ignore[no-any-return]
+    elif op == ">=":
+        return v1 >= v2  # type: ignore[no-any-return]
+    elif op == "^":
+        return v1.major == v2.major  # type: ignore[no-any-return]
+    elif op == "~":
+        return v1.major == v2.major and v1.minor == v2.minor  # type: ignore[no-any-return]
+    else:
+        raise ValueError(f"Op not supported by sem_ver: {op}")
```

Comment:
Do you know how throwing will behave here? In the JS implementation, we return `false` if none of the operators match.

Reply:
Looks like openfeature client catches exceptions well: https://github.com/open-feature/python-sdk/blob/main/openfeature/client.py#L357-L389

WDYT about returning  open feature `ParseError` instead for better error codes? https://github.com/open-feature/python-sdk/blob/main/openfeature/exception.py#L89 (would then look to do similar in other places if so)

Reply:
Ideally, this is caught by the JSON schema validator. If it's not, we try and never throw in JsonLogic and instead return a falsy value. This aligns with the out-of-the-box JsonSchema operators and provides consistency across implementations.

Reply:
Updated all the custom ops to return `None` and only log errors